RCL overwriting error state after rmw_create_node

## Bug report

**Required Info:**

- Operating System: WebAssembly
- Installation type: Source
- Version or commit hash: Galactic
- DDS implementation: Custom
- Client library (if applicable): rclcpp

#### Expected behavior
RCL should fail gracefully when `rmw_create_node` returns `NULL` and sets the error state via `RMW_SET_ERROR_MSG`

#### Actual behavior
RCL ignores the fact the `rmw_create_node` set an error and returned `NULL` and then proceeds to overwrite the error state as follows:
```
failed to initialize rcl node: rcl node's rmw handle is invalid, at /home/developer/workspace/src/ros2/rcl/rcl/src/rcl/node.c:414
>>> [rcutils|error_handling.c:108] rcutils_set_error_state()
This error state is being overwritten:
'RMW_ERROR_MESSAGE, at RMW_FILENAME:0, at /home/developer/workspace/src/ros2/rcl/rcl/src/rcl/node.c:262'
with this new error message:
'rcl node's rmw handle is invalid, at /home/developer/workspace/src/ros2/rcl/rcl/src/rcl/node.c:414'
rcutils_reset_error() should be called after error handling to avoid this.
<<<

[ERROR] [0000000003.694284999] [rcl]: Failed to fini publisher for node: 1
```

--- a/rcl/src/rcl/node.c
+++ b/rcl/src/rcl/node.c
@@ -246,13 +246,21 @@ rcl_node_init(
   node->impl->rmw_node_handle = rmw_create_node(
     &(node->context->impl->rmw_context),
     name, local_namespace_);
-  RCL_CHECK_FOR_NULL_WITH_MSG(
-    node->impl->rmw_node_handle, rmw_get_error_string().str, ret = RCL_RET_ERROR; goto fail);
+  if (NULL == node->impl->rmw_node_handle) {
+    // Preserve the error set by rmw_create_node rather than overwriting it.
+    RCL_EXPECT_ERROR_IS_SET(RCL_RET_ERROR);
+    ret = RCL_RET_ERROR;
+    goto fail;
+  }
 
   // graph guard condition
   rmw_graph_guard_condition = rmw_node_get_graph_guard_condition(node->impl->rmw_node_handle);
-  RCL_CHECK_FOR_NULL_WITH_MSG(
-    rmw_graph_guard_condition, rmw_get_error_string().str, ret = RCL_RET_ERROR; goto fail);
+  if (NULL == rmw_graph_guard_condition) {
+    // Preserve the error set by rmw_node_get_graph_guard_condition.
+    RCL_EXPECT_ERROR_IS_SET(RCL_RET_ERROR);
+    ret = RCL_RET_ERROR;
+    goto fail;
+  }
 
   node->impl->graph_guard_condition = (rcl_guard_condition_t *)allocator->allocate(
     sizeof(rcl_guard_condition_t), allocator->state);

--- a/rcl/test/rcl/test_node.cpp
+++ b/rcl/test/rcl/test_node.cpp
@@ -20,6 +20,7 @@
 
 #include "rcl/rcl.h"
 #include "rcl/node.h"
+#include "rmw/error_handling.h"
 #include "rmw/rmw.h"  // For rmw_get_implementation_identifier.
 #include "rmw/validate_namespace.h"
 #include "rmw/validate_node_name.h"
@@ -489,6 +490,26 @@ TEST_F(TestNodeFixture, test_rcl_node_init_with_internal_errors) {
     auto mock = mocking_utils::patch_and_return("lib:rcl", rmw_create_node, nullptr);
     ret = rcl_node_init(&node, name, namespace_, &context, &options);
     EXPECT_EQ(RCL_RET_ERROR, ret);
+    rcl_reset_error();
+  }
+
+  // Verify rmw_create_node error message is preserved through rcl_node_init failure path.
+  // Regression test for https://github.com/ros2/rcl/issues/983
+  {
+    auto mock = mocking_utils::patch(
+      "lib:rcl", rmw_create_node,
+      [](auto, auto, auto) -> rmw_node_t * {
+        RMW_SET_ERROR_MSG("test_rmw_create_node_failure_reason");
+        return nullptr;
+      });
+    ret = rcl_node_init(&node, name, namespace_, &context, &options);
+    EXPECT_EQ(RCL_RET_ERROR, ret);
+    ASSERT_TRUE(rcl_error_is_set());
+    // The RMW error message must be preserved, not overwritten with
+    // the generic "rcl node's rmw handle is invalid" message.
+    EXPECT_NE(
+      std::string(rcl_get_error_string().str).find("test_rmw_create_node_failure_reason"),
+      std::string::npos) << "RMW error was overwritten: " << rcl_get_error_string().str;
     rcl_reset_error();
   }
 

